src/gtk/webview_webkit2.cpp

```diff
@@ -0,0 +1,1285 @@
+    g_object_unref(result);
+
+    return can_execute;
+}
+
+bool wxWebViewWebKit::CanCut() const
+{
+    return CanExecuteEditingCommand(WEBKIT_EDITING_COMMAND_CUT);
+}
+
+bool wxWebViewWebKit::CanCopy() const
+{
+    return CanExecuteEditingCommand(WEBKIT_EDITING_COMMAND_COPY);
+}
+
+bool wxWebViewWebKit::CanPaste() const
+{
+    return CanExecuteEditingCommand(WEBKIT_EDITING_COMMAND_PASTE);
+}
+
+void wxWebViewWebKit::Cut()
+{
+    webkit_web_view_execute_editing_command(m_web_view,
+                                            WEBKIT_EDITING_COMMAND_CUT);
+}
+
+void wxWebViewWebKit::Copy()
+{
+    webkit_web_view_execute_editing_command(m_web_view,
+                                            WEBKIT_EDITING_COMMAND_COPY);
+}
+
+void wxWebViewWebKit::Paste()
+{
+    webkit_web_view_execute_editing_command(m_web_view,
+                                            WEBKIT_EDITING_COMMAND_PASTE);
+}
+
+bool wxWebViewWebKit::CanUndo() const
+{
+    return CanExecuteEditingCommand(WEBKIT_EDITING_COMMAND_UNDO);
+}
+
+bool wxWebViewWebKit::CanRedo() const
+{
+    return CanExecuteEditingCommand(WEBKIT_EDITING_COMMAND_REDO);
+}
+
+void wxWebViewWebKit::Undo()
+{
+    webkit_web_view_execute_editing_command(m_web_view,
+                                            WEBKIT_EDITING_COMMAND_UNDO);
+}
+
+void wxWebViewWebKit::Redo()
+{
+    webkit_web_view_execute_editing_command(m_web_view,
+                                            WEBKIT_EDITING_COMMAND_REDO);
+}
+
+wxString wxWebViewWebKit::GetCurrentURL() const
+{
+    // FIXME: check which encoding the web kit control uses instead of
+    // assuming UTF8 (here and elsewhere too)
+    return wxString::FromUTF8(webkit_web_view_get_uri(m_web_view));
+}
+
+
+wxString wxWebViewWebKit::GetCurrentTitle() const
+{
+    return wxString::FromUTF8(webkit_web_view_get_title(m_web_view));
+}
+
+
+static void wxgtk_web_resource_get_data_cb(WebKitWebResource *,
+                                           GAsyncResult *res,
+                                           GAsyncResult **res_out)
+{
+    *res_out = (GAsyncResult*)g_object_ref(res);
+}
+
+wxString wxWebViewWebKit::GetPageSource() const
+{
+    WebKitWebResource *resource = webkit_web_view_get_main_resource(m_web_view);
+    if (!resource)
+    {
+        return wxString("");
+    }
+
+    GAsyncResult *result = NULL;
+    webkit_web_resource_get_data(resource, NULL,
+                                 (GAsyncReadyCallback)wxgtk_web_resource_get_data_cb,
+                                 &result);
+
+    GMainContext *main_context = g_main_context_get_thread_default();
+    while (!result)
+    {
+        g_main_context_iteration(main_context, TRUE);
+    }
+
+    guchar *source = webkit_web_resource_get_data_finish(resource, result,
+                                                         NULL, NULL);
+    if (result)
+    {
+        g_object_unref(result);
+    }
+
+    if (source)
+    {
+        wxString wxs = wxString(source, wxConvUTF8);
```

Comment:
Yes, but I need to check whether source is non-NULL in order to free it, so I think I need to keep this as-is.